File: model/src/main/java/com/bnorm/barkeep/model/Bar.java
package com.bnorm.barkeep.model;

import java.util.Set;

import javax.annotation.Nullable;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonView;

@JsonInclude(JsonInclude.Include.NON_EMPTY)
public interface Bar extends HasId, Comparable<Bar> {

  @JsonView({Bar.class, User.class})
  @Nullable
  String getTitle();

  @JsonView({Bar.class, User.class})
  @Nullable
  String getDescription();

  @JsonView({Bar.class})
  @Nullable
  User getOwner();

  @JsonView({Bar.class})
  @Nullable
  Set<Ingredient> getIngredients();

  @Override
  default int compareTo(Bar o) {
    return COMPARATOR.compare(this, o);
  }
}

File: model/src/main/kotlin/com/bnorm/barkeep/model/Bar.kt
// Copyright 2017 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.model

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonView

interface BarSpec {

  @get:JsonView(Bar::class, User::class)
  val title: String?

  @get:JsonView(Bar::class, User::class)
  val description: String?

  @get:JsonView(Bar::class)
  val owner: User?
}

@JsonInclude(JsonInclude.Include.NON_EMPTY)
interface Bar : BarSpec, HasId, Comparable<Bar> {

  @get:JsonView(Bar::class)
  val ingredients: Set<Ingredient>?

  override fun compareTo(other: Bar): Int {
    return HasId.COMPARATOR.compare(this, other)
  }
}
